src/common/dout_fmt.h

```diff
@@ -0,0 +1,56 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab ft=cpp
+
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright contributors to the Ceph project
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ *
+ */
+
+#pragma once
+
+#include <iosfwd>
+#include <iterator>
+#include <fmt/core.h>
+#include "dout.h"
+
+/// \file dout_fmt.h
+///
+/// \brief dout macros to format log statements with libfmt
+///
+/// A set of dout macros taking a format string and its corresponding argument
+/// list. Log output is written directly to the underlying std::ostream by
+/// fmt::format_to() rather than exposing the stream for ostream operator
+/// chaining.
+
+// work around "warning: value computed is not used" with default dout_prefix
+inline void dout_fmt_use_prefix(std::ostream&) {}
+
+#define lsubdout_fmt(cct, sub, v, ...) \
+  dout_impl(cct, ceph_subsys_##sub, v) \
+  dout_fmt_use_prefix(dout_prefix); \
+  fmt::format_to(std::ostream_iterator<char>(*_dout), __VA_ARGS__); \
+  *_dout << dendl
+
+#define ldout_fmt(cct, v, ...) \
+  dout_impl(cct, dout_subsys, v) \
+  dout_fmt_use_prefix(dout_prefix); \
+  fmt::format_to(std::ostream_iterator<char>(*_dout), __VA_ARGS__); \
+  *_dout << dendl
+
+#define dout_fmt(v, ...) \
+  ldout_fmt((dout_context), v, __VA_ARGS__)
+
+#define ldpp_dout_fmt(dpp, v, ...) \
+  if (decltype(auto) pdpp = (dpp); pdpp) { /* workaround -Wnonnull-compare for 'this' */ \
+    dout_impl(pdpp->get_cct(), ceph::dout::need_dynamic(pdpp->get_subsys()), v) \
+    pdpp->gen_prefix(*_dout); \
+    fmt::format_to(std::ostream_iterator<char>(*_dout), __VA_ARGS__); \
```

Comment:
@cbodley If I want to mix the use of `fmt::format_to` and normal `<<` dumps, should I just be calling `fmt::format_to` manually? Maybe we can enable something like:

```
dout(20) << df("{}: {}: ", __FILE__, __FUNC__) << ... << dendl;
```

Reply:
i like the syntax. `df()` would need to return a struct that captures its arguments for the ostream operator, but i'm not sure how to preserve the constexpr-ness of the format string. this is what i came up with: https://godbolt.org/z/7sdqTMddh but it only compiles if i wrap the format string in `fmt::runtime(f.fmt)` for the call to `fmt::format_to()`. that gives up compile-time checking and makes it slower at runtime. maybe something like `df<"{}: {}: ">(__FILE__, __FUNC__)` could work, but it doesn't look like as nice

the alternative would be to use a separate statement to call format_to:
```c++
dout(2);
fmt::format_to(std::ostream_iterator<char>(*_dout), "{}: {}: ", __FILE__, __FUNC__);
*_dout << ... << dendl;
```
but it doesn't seem necessary to mix the two this way. if you want to use fmt, why can't it handle the `<< ... <<` part too?

Reply:
> @cbodley If I want to mix the use of `fmt::format_to` and normal `<<` dumps, should I just be calling `fmt::format_to` manually? Maybe we can enable something like:
> 
> ```
> dout(20) << df("{}: {}: ", __FILE__, __FUNC__) << ... << dendl;
> ```

i'd avoid this. as Casey puts, it's broken in two ways.

- it sacrifices the compile-time format check
- it materializes the format arguments even when they are not going to be formatted at all

probably we could go with:

```c++
dout_fmt(20, "{}: {}: {}", __FILE__, __FUNC__, fmt::streamed(foo));
```

Reply:
> i like the syntax. `df()` would need to return a struct that captures its arguments for the ostream operator, but i'm not sure how to preserve the constexpr-ness of the format string.

```
#define df(...) (fmt::print(*_dout, __VA_ARGS__), *_dout)
```

would work?

Reply:
neat. does that still respect the order of operator<<s? ie `"a" << df(...) << "b"` prints after a and before b?

Reply:
Yes, that's a good question. I believe that's venturing into undefined territory, nevermind.